lgc/state/PalMetadata.cpp

```diff
@@ -969,3 +920,178 @@ void PalMetadata::eraseFragmentInputInfo() {
   if (array2It != m_pipelineNode.end())
     m_pipelineNode.erase(array2It);
 }
+
+// =====================================================================================================================
+// Returns the calling convention of the first shader stage that will be executed in the pipeline.
+unsigned PalMetadata::getCallingConvenctionForFirstHardwareShaderStage() {
+  constexpr unsigned numberOfShaderStages = 5;
+  static const std::pair<unsigned, unsigned> shaderTable[numberOfShaderStages] = {
+      {mmSPI_SHADER_PGM_RSRC1_LS, CallingConv::AMDGPU_LS},
+      {mmSPI_SHADER_PGM_RSRC1_HS, CallingConv::AMDGPU_HS},
+      {mmSPI_SHADER_PGM_RSRC1_ES, CallingConv::AMDGPU_ES},
+      {mmSPI_SHADER_PGM_RSRC1_GS, CallingConv::AMDGPU_GS},
+      {mmSPI_SHADER_PGM_RSRC1_VS, CallingConv::AMDGPU_VS}};
+
+  for (unsigned i = 0; i < numberOfShaderStages; ++i) {
+    auto entry = m_registers.find(m_document->getNode(shaderTable[i].first));
+    if (entry != m_registers.end())
+      return shaderTable[i].second;
+  }
+  return CallingConv::AMDGPU_CS;
+}
+
+// =====================================================================================================================
+// Returns the offset of the first user data register for the given calling convention.
+//
+// @param callingConv : The calling convention
+unsigned PalMetadata::getFirstUserDataReg(unsigned callingConv) {
+  static const std::map<unsigned, unsigned> shaderTable = {{CallingConv::AMDGPU_LS, mmSPI_SHADER_USER_DATA_LS_0},
+                                                           {CallingConv::AMDGPU_HS, mmSPI_SHADER_USER_DATA_HS_0},
+                                                           {CallingConv::AMDGPU_ES, mmSPI_SHADER_USER_DATA_ES_0},
+                                                           {CallingConv::AMDGPU_GS, mmSPI_SHADER_USER_DATA_GS_0},
+                                                           {CallingConv::AMDGPU_VS, mmSPI_SHADER_USER_DATA_VS_0}};
+  static const std::map<unsigned, unsigned> shaderTableGfx9 = {{CallingConv::AMDGPU_LS, mmSPI_SHADER_USER_DATA_LS_0},
```

Comment:
I guess you need this separate table/map because of merged es-gs using mmSPI_SHADER_USER_DATA_ES_0 on gfx9. (Was that wrong in my code? Is that how you ended up here as part of your GS work?) But I think the gfx9 table does not need entries for LS or ES at all.

Also same comment as before about using a static map.

Reply:
Yes that is correct.  That was wrong in the original code, and I need to fix is before I can enable reloc geometry pipelines.  I've removed those entries.

Reply:
I added them back in. I need to size of the two arrays to match make the assignment to the arrayref to work.